python/semantic_kernel/template_engine/template_tokenizer.py

```diff
@@ -0,0 +1,180 @@
+# Copyright (c) Microsoft. All rights reserved.
+
+from logging import Logger
+from typing import List
+
+from semantic_kernel.template_engine.blocks.block import Block
+from semantic_kernel.template_engine.blocks.block_types import BlockTypes
+from semantic_kernel.template_engine.blocks.code_block import CodeBlock
+from semantic_kernel.template_engine.blocks.symbols import Symbols
+from semantic_kernel.template_engine.blocks.text_block import TextBlock
+from semantic_kernel.template_engine.code_tokenizer import CodeTokenizer
+from semantic_kernel.utils.null_logger import NullLogger
+
+
+# BNF parsed by TemplateTokenizer:
+# [template]       ::= "" | [block] | [block] [template]
+# [block]          ::= [sk-block] | [text-block]
+# [sk-block]       ::= "{{" [variable] "}}"
+#                      | "{{" [value] "}}"
+#                      | "{{" [function-call] "}}"
+# [text-block]     ::= [any-char] | [any-char] [text-block]
+# [any-char]       ::= any char
+class TemplateTokenizer:
+    def __init__(self, log: Logger = None):
+        self.log = log or NullLogger()
+        self.code_tokenizer = CodeTokenizer(self.log)
+
+    def tokenize(self, text: str) -> List[Block]:
+        # An empty block consists of 4 chars: "{{}}"
+        EMPTY_CODE_BLOCK_LENGTH = 4
+        # A block shorter than 5 chars is either empty or
+        # invalid, e.g. "{{ }}" and "{{$}}"
+        MIN_CODE_BLOCK_LENGTH = EMPTY_CODE_BLOCK_LENGTH + 1
+
+        text = text or ""
+
+        # Render None/empty to ""
+        if not text or text == "":
+            return [TextBlock("", log=self.log)]
+
+        # If the template is "empty" return it as a text block
+        if len(text) < MIN_CODE_BLOCK_LENGTH:
+            return [TextBlock(text, log=self.log)]
+
+        blocks = []
+        end_of_last_block = 0
+        block_start_pos = 0
+        block_start_found = False
+        inside_text_value = False
+        text_value_delimiter = None
+        skip_next_char = False
+        next_char = text[0]
+
+        for next_char_cursor in range(1, len(text)):
+            current_char_pos = next_char_cursor - 1
+            cursor = next_char_cursor
+            current_char = next_char
+            next_char = text[next_char_cursor]
+
+            if skip_next_char:
+                skip_next_char = False
+                continue
+
+            # When "{{" is found outside a value
+            # Note: "{{ {{x}}" => ["{{ ", "{{x}}"]
+            if (
+                not inside_text_value
+                and current_char == Symbols.BLOCK_STARTER
+                and next_char == Symbols.BLOCK_STARTER
+            ):
+                # A bock starts at the first "{"
```

Comment:
Small Typo, should be:
`# A block starts at the first "{"`

Reply:
Thanks, fixed!